crates/sui-framework/packages/sui-framework/sources/kiosk/kiosk.move

```diff
@@ -412,6 +415,181 @@ module sui::kiosk {
+        /// owner can't delete the `Bag` storage unless the extension provides
+        /// the functionality, they can revoke the permissions at any time,
+        /// making the extension a harmless storage - similar to Kiosk having an
+        /// exposed UID.
+        is_enabled: bool,
+    }
+
+    /// The `ExtensionKey` is a typed dynamic field key used to store the
+    /// extension configuration and data. `Ext` is a phantom type that is used
+    /// to identify the extension witness.
+    struct ExtensionKey<phantom Ext> has store, copy, drop {}
+
+    /// Add an extension to the Kiosk. Can only be performed by the owner.
+    ///
+    /// Unlike the original implementation, this one uses the `public`
+    /// visibility modifier to allow "owned kiosk" extension. Ideally we don't
+    /// want this function to be called arbitrarily, and to prevent some
+    /// malicious scenarios we now require the extension witness on install.
+    public fun add_extension<Ext: drop>(
+        _ext: Ext,
+        self: &mut Kiosk,
+        cap: &KioskOwnerCap,
+        permissions: u32,
+        types: vector<TypeName>,
+        ctx: &mut TxContext
+    ) {
+        assert!(object::id(self) == cap.for, ENotOwner);
+
+        df::add(&mut self.id, ExtensionKey<Ext> {}, Extension {
+            storage: bag::new(ctx),
+            permissions,
+            types,
+            is_enabled: true,
+        })
+    }
+
+    /// Revoke permissions from the extension. While it does not remove the
+    /// extension completely, it keeps it from performing any protected actions.
+    /// The storage is still available to the extension (until it's removed).
+    public fun disable_extension<Ext: drop>(
+        self: &mut Kiosk,
+        cap: &KioskOwnerCap,
+    ) {
+        assert!(object::id(self) == cap.for, ENotOwner);
+        extension_mut<Ext>(self).is_enabled = false;
+    }
+
+    /// Re-enable permissions for the extension. Can only be performed by the
+    /// owner. The extension can start performing protected actions again.
+    public fun re_enable_extension<Ext: drop>(
+        self: &mut Kiosk,
+        cap: &KioskOwnerCap,
+    ) {
+        assert!(object::id(self) == cap.for, ENotOwner);
+        extension_mut<Ext>(self).is_enabled = true;
+    }
+
+    /// Limit the extension to a set of types. Can only be performed by the
+    /// owner. The extension can only perform protected actions on the types
+    /// specified in the list.
+    public fun limit_extension_for_types<Ext: drop>(
+        self: &mut Kiosk,
+        cap: &KioskOwnerCap,
+        types: vector<TypeName>,
+    ) {
+        assert!(object::id(self) == cap.for, ENotOwner);
+        extension_mut<Ext>(self).types = types;
+    }
+
+    /// Get immutable access to the extension storage. Can only be performed by
+    /// the extension as long as the extension is installed.
+    public fun ext_storage<Ext: drop>(
```

Comment:
Having so many functions named `ext_` is making me wonder whether we want to have this in a separate `extensions` module rather than co-locating it with `kiosk`

Reply:
All attempts to move it outside ended nowhere, access control is very tight in Kiosk.